src/db/gc/mod.rs

```diff
@@ -0,0 +1,217 @@
+// Copyright 2019-2023 ChainSafe Systems
+// SPDX-License-Identifier: Apache-2.0, MIT
+
+//!
+//! The current implementation of the garbage collector is concurrent mark-and-sweep.
+//!
+//! ## Design goals
+//! A correct GC algorithm that is simple and efficient for forest scenarios.
+//!
+//! ## GC Algorithm
+//! The `mark-and-sweep` algorithm was chosen due to it's simplicity, efficiency and low memory
+//! footprint. We do have to generate keys from values for some of the iterated columns due to a
+//! ParityDB limitation. See <https://github.com/paritytech/parity-db/issues/187>.
+//! Previously the `semi-space` algorithm was used resulting in data duplication and up to 100%
+//! extra disk usage.
+//!
+//! ## GC Workflow
+//! 1. Mark: traverse all the relevant database columns, generating integer hash representations for
+//! each database key and storing those in a [`HashSet`].
+//! 2. Wait at least `chain finality` blocks.
+//! 3. Traverse reachable blocks starting from the current heaviest tipset and remove those from the
+//! marked `HashSet`, leaving only unreachable entries that are older than `chain finality` to avoid
+//! removing something that could later become reachable as a result of a fork.
+//! 4. Sweep, removing all the remaining marked entries from the database.
+//!
+//! ## Correctness
+//! This algorithm is traversing the reachable graph using the same tooling as the snapshot export.
+//! Therefore it ensures that we eliminate all the reachable from the set scheduled for removal.
+//! Additionally, it waits at least `chain finality` before filtering out and sweeping marked
+//! records, making sure nothing that could have become reachable in the meantime gets removed.
+//! A snapshot can be used to bootstrap the node from scratch, therefore the algorithm is considered
+//! correct when a valid snapshot can be exported using records available in the database after
+//! sweeping.
+//!
+//! ## Disk usage
+//! There's no additional disk space required to run this algorithm.
```

Comment:
Well, there's more to be said here. Will 10GiB of live data take up more than 10GiB of disk space? Are all unreachable keys guaranteed to be garbage collected?